src/simpleMask.js

```diff
@@ -0,0 +1,171 @@
+
+    const maskMethods = {
+        reverse: 'reverse',
+        forward: 'forward'
+    }
+
+    const availableCharacterPatterns =  [settings.numeric, settings.alphanumeric, !!settings.customPattern]
+    const requiredParams = [containerEl, settings.inputSelector, settings.format]
+
+
+    let inputEl;
+    let output;
+    let builtValue;
+    let formatArrayOfCharacters;
+    let arrayOfSpecialCharactersInFormat;
+
+
+    const init = () => {
+        if (settingsAreValid()) {
+            inputEl = containerEl.querySelector(settings.inputSelector);
+            initPatternData();
+            bindEvents();
+        } else {
+            throw new Error('Missing one or many of the following: containerEl, input selector, numeric, alphanumeric or custom pattern')
+        }
+
+    };
+
+    const settingsAreValid = () => {
+        const selectedCharacterPattern = availableCharacterPatterns.filter(pattern => pattern === true);
+        const requiredParamsArePresent = requiredParams.every(param => param);
+
+        let settingsAreValid = true;
+
+        if(selectedCharacterPattern.length !== 1 || !requiredParamsArePresent) {
+            settingsAreValid = false;
+        }
+
+        return settingsAreValid;
+    }
+
+    const initPatternData = () => {
+        formatArrayOfCharacters = settings.format.split('');
+        arrayOfSpecialCharactersInFormat = getSpecialCharacters(formatArrayOfCharacters);
+    }
+
+    const getSpecialCharacters = arrayOfCharacters => {
+        let object = [];
+
+        arrayOfCharacters.forEach((char, index) => {
+            if (char !== settings.maskPlaceholder) {
+                object.push({ char, index });
+            }
+        });
+
+        return object;
+    };
+
+    const bindEvents = () => {
+        inputEl.addEventListener('focusin', maskInput);
+        inputEl.addEventListener('focusout', maskInput);
+        inputEl.addEventListener('keypress', checkForInvalidKey);
+    };
+
+    const maskInput = (e) => {
+        if (inputEl.value.length !== 0 && inputEl.value.length >= getFormatLength() ) {
+            const value = getMaskValue(arrayOfSpecialCharactersInFormat, getMaskMethod(e));
+            setInputValue(value);
+
+            if (settings.onMaskCallback) {
+                settings.onMaskCallback(value)
+            }
+        }
+    };
+
+    const getMaskMethod = (e) => e.type === 'focusin' ? maskMethods.reverse : maskMethods.forward;
+
+    const setInputValue = (value) => inputEl.value = value;
+
+    const getMaskValue = (arrayOfSpecialCharacters, maskMethod) => {
+        let maskedValue;
+
+        arrayOfSpecialCharacters.forEach(specialCharObject => {
+            if (specialCharObject.char) {
+                maskedValue = buildMaskedValue(inputEl.value, specialCharObject, maskMethod);
+            }
+        });
+
+        builtValue = null;
+        return maskedValue;
+    };
+
+    const buildMaskedValue = (inputValue, specialCharObject, maskMethod) => {
```

Comment:
[question] I'm a little confused as to how builtValue works in this function and `getMaskValue`. Is it a variable that is only needed in these two functions? Would it make sense to remove builtValue from the plugin scope and make the `forEach` on line 98 an `reduce`? Then you could pass the accumulator (the buildValue replacement here) into buildMaskedValue()?

Reply:
Very good call here. This was probably the most confusing part for me. I can't remember the last time I needed `reduce`, so I didn't think about it! It's updated now.